proof/aes.cry

```diff
@@ -0,0 +1,240 @@
+// GF28 operations
+gf28Add : {n} (fin n) => [n]GF28 -> GF28
+gf28Add ps = sums ! 0
+  where sums = [zero] # [  p ^ s | p <- ps | s <- sums ]
+
+irreducible = <| x^^8 + x^^4 + x^^3 + x + 1 |>
+
+gf28Mult : (GF28, GF28) -> GF28
+gf28Mult (x, y) = pmod(pmult x y) irreducible
+
+gf28Pow : (GF28, [8]) -> GF28
+gf28Pow (n, k) = pow k
+  where   sq x  = gf28Mult (x, x)
+          odd x = x ! 0
+          pow i = if i == 0 then 1
+                  else if odd i
+                       then gf28Mult(n, sq (pow (i >> 1)))
+                       else sq (pow (i >> 1))
+
+gf28Inverse : GF28 -> GF28
+gf28Inverse x = gf28Pow (x, 254)
+
+gf28DotProduct : {n} (fin n) => ([n]GF28, [n]GF28) -> GF28
+gf28DotProduct (xs, ys) = gf28Add [ gf28Mult (x, y) | x <- xs
+                                                    | y <- ys ]
+
+gf28VectorMult : {n, m} (fin n) => ([n]GF28, [m][n]GF28) -> [m]GF28
+gf28VectorMult (v, ms) = [ gf28DotProduct(v, m) | m <- ms ]
+
+gf28MatrixMult : {n, m, k} (fin m) => ([n][m]GF28, [m][k]GF28) -> [n][k]GF28
+gf28MatrixMult (xss, yss) = [ gf28VectorMult(xs, yss') | xs <- xss ]
+   where yss' = transpose yss
+
+// The affine transform and its inverse
+xformByte : GF28 -> GF28
+xformByte b = gf28Add [b, (b >>> 4), (b >>> 5), (b >>> 6), (b >>> 7), c]
+   where c = 0x63
+
+xformByte' : GF28 -> GF28
+xformByte' b = gf28Add [(b >>> 2), (b >>> 5), (b >>> 7), d] where d = 0x05
+
+// The SubBytes transform and its inverse
+SubByte : GF28 -> GF28
+SubByte b = xformByte (gf28Inverse b)
+
+SubByte' : GF28 -> GF28
+SubByte' b = sbox@b
+
+SubBytes : State -> State
+SubBytes state = [ [ SubByte' b | b <- row ] | row <- state ]
+
+
+InvSubByte : GF28 -> GF28
+InvSubByte b = gf28Inverse (xformByte' b)
+
+InvSubBytes : State -> State
+InvSubBytes state = [ [ InvSubByte b | b <- row ] | row <- state ]
+
+// The ShiftRows transform and its inverse
+ShiftRows : State -> State
+ShiftRows state = [ row <<< shiftAmount | row <- state
+                                        | shiftAmount <- [0 .. 3]
+                  ]
+
+InvShiftRows : State -> State
+InvShiftRows state = [ row >>> shiftAmount | row <- state
+                                           | shiftAmount <- [0 .. 3]
+                     ]
+
+// The MixColumns transform and its inverse
+MixColumns : State -> State
+MixColumns state = gf28MatrixMult (m, state)
+    where m = [[2, 3, 1, 1],
+               [1, 2, 3, 1],
+               [1, 1, 2, 3],
+               [3, 1, 1, 2]]
+
+InvMixColumns : State -> State
+InvMixColumns state = gf28MatrixMult (m, state)
+    where m = [[0x0e, 0x0b, 0x0d, 0x09],
+               [0x09, 0x0e, 0x0b, 0x0d],
+               [0x0d, 0x09, 0x0e, 0x0b],
+               [0x0b, 0x0d, 0x09, 0x0e]]
+
+// The AddRoundKey transform
+AddRoundKey : (RoundKey, State) -> State
+AddRoundKey (rk, s) = rk ^ s
+// Key expansion
+Rcon : [8] -> [4]GF28
+Rcon i = [(gf28Pow (<| x |>, i-1)), 0, 0, 0]
+
+SubWord : [4]GF28 -> [4]GF28
+SubWord bs = [ SubByte b | b <- bs ]
```

Comment:
I was looking at https://github.com/GaloisInc/cryptol/blob/master/examples/AES.cry, I think you just have an older copy of that.

I actually asked about aes.cry and aes.saw, and that they don't seem to be used to test anything in OpenSSL. So why did they get added? Would it be useful to actually use them for something?